Framework/Geometry/src/Objects/MeshObject.cpp

```diff
@@ -0,0 +1,749 @@
+  const double scalTripProd = ao.scalar_prod(bo.cross_prod(co));
+  const double denom =
+      modao * modbo * modco + modco * aobo + modbo * aoco + modao * boco;
+  if (denom != 0.0)
+    return 2.0 * atan2(scalTripProd, denom);
+  else
+    return 0.0; // not certain this is correct
+}
+
+/**
+* Calculate if a point PT is a valid point on the track
+* @param Pt :: Point to calculate from.
+* @param uVec :: Unit vector of the track
+* @retval 0 :: Not valid / double valid
+* @retval 1 :: Entry point
+* @retval -1 :: Exit Point
+*/
+int MeshObject::calcValidType(const Kernel::V3D &Pt,
+                              const Kernel::V3D &uVec) const {
+  const Kernel::V3D shift(uVec * Kernel::Tolerance * 25.0);
+  const Kernel::V3D testA(Pt - shift);
+  const Kernel::V3D testB(Pt + shift);
+  const int flagA = isValid(testA);
+  const int flagB = isValid(testB);
+  if (!(flagA ^ flagB))
+    return 0;
+  return (flagA) ? -1 : 1;
+}
+
+/**
+* Takes input axis aligned bounding box max and min points and calculates the
+*bounding box for the
+* object and returns them back in max and min points.
+*
+* @param xmax :: Maximum value for the bounding box in x direction
+* @param ymax :: Maximum value for the bounding box in y direction
+* @param zmax :: Maximum value for the bounding box in z direction
+* @param xmin :: Minimum value for the bounding box in x direction
+* @param ymin :: Minimum value for the bounding box in y direction
+* @param zmin :: Minimum value for the bounding box in z direction
+*/
+void MeshObject::getBoundingBox(double &xmax, double &ymax, double &zmax,
+                                double &xmin, double &ymin,
+                                double &zmin) const {
+  BoundingBox bb = getBoundingBox();
+  xmin = bb.xMin();
+  xmax = bb.xMax();
+  ymin = bb.yMin();
+  ymax = bb.yMax();
+  zmin = bb.zMin();
+  zmax = bb.zMax();
+
+  return;
+}
+
+/**
+* Find solid angle of object wrt the observer.
+* @param observer :: point to measure solid angle from
+* @return :: estimate of solid angle of object.
+*/
+double MeshObject::solidAngle(const Kernel::V3D &observer) const {
+
+  double solidAngleSum(0), solidAngleNegativeSum(0);
+  V3D vertex1, vertex2, vertex3;
+  for (size_t i = 0; getTriangle(i, vertex1, vertex2, vertex3); ++i) {
+    double sa = getTriangleSolidAngle(vertex1, vertex2, vertex3, observer);
+    if (sa > 0.0) {
+      solidAngleSum += sa;
+    } else {
+      solidAngleNegativeSum += sa;
+    }
+  }
+  return 0.5 * (solidAngleSum - solidAngleNegativeSum);
+}
+
+/**
+* Find solid angle of object wrt the observer with a scaleFactor for the object.
+* @param observer :: point to measure solid angle from
+* @param scaleFactor :: V3D giving scaling of the object
+* @return :: estimate of solid angle of object.
+*/
+double MeshObject::solidAngle(const Kernel::V3D &observer,
+                              const Kernel::V3D &scaleFactor) const
+
+{
+  MeshObject scaledObject;
+  std::vector<V3D> scaledVertices;
```

Comment:
Given the number of elements to be held in the scaledVertices vector, would it be appropriate to reserve that space before pushing to the vector in an iteration, i.e.:
scaledVertices.reserve(m_vertices.size());